src/hdmf/common/table.py

```diff
@@ -470,37 +487,39 @@ def __eq__(self, other):
             {'name': 'index', 'type': (bool, VectorIndex, 'array_data'),
              'doc': 'whether or not this column should be indexed', 'default': False})
     def add_column(self, **kwargs):
-        name = getargs('name', kwargs)
-        for col in self.__columns__:
-            if col['name'] == name:  # column has not been added but is pre-specified
-                msg = "column '%s' already exists in %s '%s'" % (name, self.__class__.__name__, self.name)
-                raise ValueError(msg)
-
-        self._add_column(**kwargs)
-
-    @docval({'name': 'name', 'type': str, 'doc': 'the name of this VectorData'},
-            {'name': 'description', 'type': str, 'doc': 'a description for this column'},
-            {'name': 'data', 'type': ('array_data', 'data'),
-             'doc': 'a dataset where the first dimension is a concatenation of multiple vectors', 'default': list()},
-            {'name': 'table', 'type': (bool, 'DynamicTable'),
-             'doc': 'whether or not this is a table region or the table the region applies to', 'default': False},
-            {'name': 'index', 'type': (bool, VectorIndex, 'array_data'),
-             'doc': 'whether or not this column should be indexed', 'default': False})
-    def _add_column(self, **kwargs):
         """
         Add a column to this table.
 
         If data is provided, it must contain the same number of rows as the current state of the table.
 
-        :raises ValueError
+        :raises ValueError: if the column has already been added to the table
         """
-        name, data = getargs('name', 'data', kwargs)
+        name, data, description = getargs('name', 'data', 'description', kwargs)
         index, table = popargs('index', 'table', kwargs)
 
         if name in self.__colids:  # column has already been added
             msg = "column '%s' already exists in %s '%s'" % (name, self.__class__.__name__, self.name)
             raise ValueError(msg)
 
+        if name in self.__uninit_cols:  # column is a predefined optional column from the spec
+            # check the given values against the predefined optional column spec. if they do not match, raise a warning
+            # and ignore the given arguments. users should not be able to override these values
+            spec_table = self.__uninit_cols[name].get('table', False)
+            if table != spec_table:
+                msg = ("Column '%s' is predefined in %s with table=%s which does not match the entered "
+                       "table argument. The entered table argument will be ignored."
```

Comment:
@rly I'm a bit wary of this different behavior. It seems like with the 1.6.1 version of `hdmf` one could use `add_column()` to overwrite an existing column. This may not have been the *intended* behavior, but since it was allowed, it is effectively part of the `feature-set` of version < 2.0.0 of `hdmf`.

I see a lot of instances in the AllenSDK where a `Table` is created and filled with some initial values and then additional columns are tacked on to the existing `Table`.

See: https://github.com/AllenInstitute/AllenSDK/blob/c0d5d40be5c48e089b7121b2dcb4557504aa2dd4/allensdk/brain_observatory/ecephys/write_nwb/__main__.py#L718

https://github.com/AllenInstitute/AllenSDK/blob/c0d5d40be5c48e089b7121b2dcb4557504aa2dd4/allensdk/brain_observatory/ecephys/write_nwb/__main__.py#L458

Is there a workaround such as deleting the existing columns if we *do* want this overwrite feature?

Reply:
Maybe adding an additional default parameter to `add_column()` like `overwrite_existing=False`?

Reply:
@njmei You raise a valid concern. However, let's take the `Units` table and the optional `spike_times` dataset as an example. The `spike_times` column is specified in the NWB spec as an optional, ragged array column. The `Units` class predefines the `spike_times` column but does not add it to the `Units` table until the user either adds the column manually or adds a row with a value for `spike_times`. The check in the code here is to ensure that if the user adds the `spike_times` column, that certain column properties match the predefined column in the `Units` class, which conforms with the NWB spec. If the user is allowed to add the `spike_times` column but make it not a ragged array or make it a table region reference, then the resulting file will not be compliant with the NWB spec. 

This change does break existing functionality, but this was uncaught buggy functionality -- HDMF should not allow users to write files that are non-compliant with the spec. As long as the code that uses `add_column` does not use it to create non-compliant files, it should not be affected. Let me know if you find that not to be the case. @ajtritt @oruebel feel free to chime in.

Reply:
I decided to revert the behavior and allow overriding of predefined columns with a warning that the user does this at their own risk. It is more dangerous to ignore the passed in `table` or `index` arguments because the `data` rely on those arguments and will not make sense anymore. 

We can make this condition a hard error in a future major version of HDMF.